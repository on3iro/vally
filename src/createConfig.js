// @flow

import type { Validator, Config } from './types'

const createConfig = (specs: Array<{
  selector: string,
  validators: Array<Validator>
}>
): Config => {
  // $FlowFixMe
  return {
    fields: specs.reduce((acc, spec) => {
      const node = document.querySelector(spec.selector)

      if (!(
        node &&
        (
          node.mock ||
          (node instanceof HTMLInputElement || node instanceof HTMLSelectElement)
        )
      )) {
        console.warn(`vally, createConfig: node with selector "${spec.selector}" could not be found!`)

        return acc
      }

      return [ ...acc, {
        node,
        validators: spec.validators
      } ]
    }, [])
  }
}

export default createConfig
