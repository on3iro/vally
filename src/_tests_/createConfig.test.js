// @flow

import {
  isNoneEmptyString,
  isNumberString
} from '../validatorFns'

import createConfig from '../createConfig'

describe('createConfig()', () => {
  const nonEmptyValidator = {
    fn: isNoneEmptyString,
    type: 'NONE_EMPTY'
  }

  const numberValidator = {
    fn: isNumberString,
    type: 'NUMBER'
  }

  const noneEmptyNode = '<input class="noneEmptyNode" value="some string">'
  const numberNode = '<input class="numberNode" value="8">'
  const emptyStrNode = '<input class="emptyStrNode" value="" required>'

  const setUp = (specOW) => {
    if (document && document.body) {
      document.body.innerHTML = `
        ${noneEmptyNode}
        ${numberNode}
        ${emptyStrNode}
      `
    }

    const spec = specOW || [{
      selector: '.noneEmptyNode',
      validators: [ nonEmptyValidator ]
    }, {
      selector: '.numberNode',
      validators: [ numberValidator ]
    }, {
      selector: '.emptyStrNode',
      validators: [ nonEmptyValidator ]
    }]

    const config = createConfig(spec)

    return {
      config
    }
  }

  it('should return valid config with nodes', () => {
    const { config } = setUp()

    expect(config.fields.length).toBe(3)
    expect(config).toMatchSnapshot()
  })

  it('should skip spec if node cannot be found', () => {
    const { config } = setUp([{
      selector: '.noneEmptyNode',
      validators: [ nonEmptyValidator ]
    }, {
      selector: '.doesNotExist',
      validators: [ nonEmptyValidator ]
    }, {
      selector: '.numberNode',
      validators: [ numberValidator ]
    }, {
      selector: '.emptyStrNode',
      validators: [ nonEmptyValidator ]
    }])

    expect(config.fields.length).toBe(3)
    expect(config).toMatchSnapshot()
  })
}) // end createConfig()
