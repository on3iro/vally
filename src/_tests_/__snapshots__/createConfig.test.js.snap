// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createConfig() should return valid config with nodes 1`] = `
Object {
  "fields": Array [
    Object {
      "node": <input
        class="noneEmptyNode"
        value="some string"
      />,
      "validators": Array [
        Object {
          "fn": [Function],
          "type": "NONE_EMPTY",
        },
      ],
    },
    Object {
      "node": <input
        class="numberNode"
        value="8"
      />,
      "validators": Array [
        Object {
          "fn": [Function],
          "type": "NUMBER",
        },
      ],
    },
    Object {
      "node": <input
        class="emptyStrNode"
        required=""
        value=""
      />,
      "validators": Array [
        Object {
          "fn": [Function],
          "type": "NONE_EMPTY",
        },
      ],
    },
  ],
}
`;

exports[`createConfig() should skip spec if node cannot be found 1`] = `
Object {
  "fields": Array [
    Object {
      "node": <input
        class="noneEmptyNode"
        value="some string"
      />,
      "validators": Array [
        Object {
          "fn": [Function],
          "type": "NONE_EMPTY",
        },
      ],
    },
    Object {
      "node": <input
        class="numberNode"
        value="8"
      />,
      "validators": Array [
        Object {
          "fn": [Function],
          "type": "NUMBER",
        },
      ],
    },
    Object {
      "node": <input
        class="emptyStrNode"
        required=""
        value=""
      />,
      "validators": Array [
        Object {
          "fn": [Function],
          "type": "NONE_EMPTY",
        },
      ],
    },
  ],
}
`;
